### PR TITLE
Make QgsRectangle.toString( int thePrecision ) callable

### PR DESCRIPTION
**Pull request by jdugge**
Previously, QgsRectangle.toString( int thePrecision ) couldn't be called in Python because all integer arguments were implicitly converted to booleans and then used to call QgsRectangle.toString( bool automaticPrecision ) instead.

### CODE REVIEW

**Comment by nyalldawson** [comment]
This API is a bit odd in general. I think a better approach would be a single toString method:
   
    /**
     * Returns a string representation of form xmin,ymin xmax,ymax. Coordinates will be truncated
     * to the specified \a precision. If \a precision is less than 0 then a suitable minimum precision
     * will be automatically calculated.
    */
    QString toString( int precision = 16 ) const;

**Comment by wonder-sk** [comment]
Agreed with Nyall, additionally I would suggest to have default precision value set to -1

**Comment by jdugge** [comment]
I think merging the two methods is a good idea.

I wouldn't make the automatic precision determination the default, though, at least not in its current implementation, because I find that the automatically rounded results can be a bit confusing.

The algorithm will round to integer values as soon as both sides of the rectangle are at least one unit long, which can lead to an output of `'-0,-0 : 0,0'` for a `1 x 1` rectangle:

```
>>> qgis.core.QgsRectangle(-0.5,-0.5,0.5,0.5).toString(automaticPrecision=True)
'-0,-0 : 0,0'
```

Using a less extreme example, I would expect the rectangle `0.4,0 : 2.6,1` to be returned unchanged, but the current implementation will output '0,0 : 3,1'.

Maybe the automatic rounding could be made slightly more conservative before merging the two `toString` methods and making automatic rounding the default?

**Comment by wonder-sk** [comment]
Hmm... I didn't realize it behaved so badly... sounds like the automatic precision detection just does not work!

**Comment by m-kuhn** [comment]
For reference, this is what I get here

```
qgis.core.QgsRectangle(-0.5,-0.5,0.5,0.5).toString(automaticPrecision=True)
'-1,-1 : 1,1'
qgis.core.QgsRectangle(-0.5,-0.5,0.5,0.5).toString(1)
'-1,-1 : 1,1'
qgis.core.QgsRectangle(-0.5,-0.5,0.5,0.5).toString(2)
'-1,-1 : 1,1'
qgis.core.QgsRectangle(-0.5,-0.5,0.5,0.5).toString(0)
'-0.5000000000000000,-0.5000000000000000 : 0.5000000000000000,0.5000000000000000'
qgis.core.QgsRectangle(-0.5,-0.5,0.5,0.5).toString(16)
'-1,-1 : 1,1'
qgis.core.QgsRectangle(-0.5,-0.5,0.5,0.5).toString(18)
'-1,-1 : 1,1'
```

**Comment by jdugge** [comment]
@m-kuhn: That's interesting that it seems to round differently for you. I get `'-1,-1 : 1,1'` both on a Windows 7 machine (using Qt 4.8.5) and on Ubuntu 16.04 (using Qt 4.8.7 and 5.5.1). What system did you run this on?

The fact that argument values of `automaticPrecision=True`, `1`, `2`, `16`, and `18` all yield the same (automatically rounded) result, while `automaticPrecision=False` and `0` yield the full precision output is the issue that's addressed by the pull request as it currently stands

**Comment by m-kuhn** [comment]
Fedora, Qt 5.7

File: src/core/qgsrectangle.h [hunk 1]
```diff
@@ -115,10 +115,10 @@ class CORE_EXPORT QgsRectangle
     QString asWktPolygon() const;
     //! returns a QRectF with same coordinates.
     QRectF toRectF() const;
-    //! returns string representation of form xmin,ymin xmax,ymax
-    QString toString( bool automaticPrecision = false ) const;
-    //! overloaded toString that allows precision of numbers to be set
-    QString toString( int thePrecision ) const;
+    //! returns string representation of form xmin,ymin : xmax,ymax
+    //! Coordinates will be truncated to the specified precision.
+    //! If the specified precision is less than 0, a suitable minimum precision is used.
```
Comment (m-kuhn) [inline src/core/qgsrectangle.h]: Multiline doxygen comments are best served as block comments

```
/**
 * X
 * Y
 * Z
 */
```

File: python/core/qgsrectangle.sip [hunk 2]
```diff
@@ -90,10 +90,12 @@ class QgsRectangle
...
-    //! returns string representation of form xmin,ymin xmax,ymax
-    QString toString( bool automaticPrecision = false ) const;
-    //! overloaded toString that allows precision of numbers to be set
-    QString toString( int precision ) const;
+    /**
+     * returns a string representation of form xmin,ymin : xmax,ymax
+     * Coordinates will be truncated to the specified precision.
+     * If the specified precision is less than 0, a suitable minimum precision is used.
+     */
+    QString toString( int thePrecision = 16 ) const;
```
Comment (m-kuhn) [inline python/core/qgsrectangle.sip]: FYI, all `the`-prefixes have been removed from the code recently.

Comment (jdugge) [inline python/core/qgsrectangle.sip]: Done (and finally managed to get the formatting right as well)

**Comment by nyalldawson** [comment]
Thanks @jdugge !

**Comment by jdugge** [comment]
@nyalldawson: Pleasure!
@m-kuhn: Thanks for your patience in reviewing this. I promise the next pull request will be smoother

**Pull request merged**